In [1]:
#Import Libraries
import pandas as pd
import numpy as numpy
import matplotlib.pyplot as pyplot
import seaborn as sns
import warnings


warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

In [2]:
# Load Instagram Data
df = pd.read_csv('./data/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Instagram.csv', low_memory=False)

In [3]:
# Check data dimensions
print('Rows: ', df.shape[0])
print('Columns: ', df.shape[1])

Rows:  10000
Columns:  147


In [4]:
# Examine first two observations
df.head(2)

Date            Post ID    Network Post Type Content Type  \
0  1/30/2023 5:06 pm  18216468940204235  Instagram      Post     Carousel   
1  4/15/2022 9:04 am  17971531720518163  Instagram      Post        Photo   

       Profile        Sent by                                      Link  \
0  stanbicibtc                 https://www.instagram.com/p/CoC6CKBIHbV/   
1  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CcXUQTEMRm7/   

                                                                                                                                                                                                                                                                           Post  \
0  The new Naira banknotes in circulation have several security features for ease of recognition which can be easily seen and touched.\nPlease deposit your old N200, N500 and N1000 naira notes at any Stanbic IBTC branch near you.\nImage Source - @centralbankng \n#ITCANBE   
1                                                                                                                                                                       We hope the faith that this #GoodFriday restores abides with you and your loved ones. \n#TGIF\n#ITCANBE   

   Linked Content Impressions Organic Impressions  Viral Impressions  \
0             NaN      39,681              39,681                NaN   
1             NaN      29,186              29,186                NaN   

   Non-viral Impressions  Paid Impressions  Fan Impressions  \
0                    NaN               NaN              NaN   
1                    NaN               NaN              NaN   

   Fan Organic Impressions  Fan Paid Impressions  Non-fan Impressions  \
0                      NaN                   NaN                  NaN   
1                      NaN                   NaN                  NaN   

   Non-fan Organic Impressions  Non-fan Paid Impressions   Reach  \
0                          NaN                       NaN  32,418   
1                          NaN                       NaN  25,055   

  Organic Reach  Viral Reach  Non-viral Reach  Paid Reach  Fan Reach  \
0        32,418          NaN              NaN         NaN        NaN   
1        25,055          NaN              NaN         NaN        NaN   

   Fan Paid Reach  Potential Reach Engagement Rate (per Impression)  \
0             NaN              NaN                            4.89%   
1             NaN              NaN                            6.04%   

  Engagement Rate (per Reach) Engagements Reactions  Likes  Dislikes  \
0                       5.99%       1,942     1,261  1,261       NaN   
1                       7.04%       1,763       709    709       NaN   

   Love Reactions  Haha Reactions  Wow Reactions  Sad Reactions  \
0             NaN             NaN            NaN            NaN   
1             NaN             NaN            NaN            NaN   

   Angry Reactions Comments  Shares  Saves  Click-Through Rate  \
0              NaN       56     NaN  625.0                 NaN   
1              NaN    1,033     NaN   21.0                 NaN   

   Post Link Clicks  Other Post Clicks  Post Clicks (All)  Post Media Clicks  \
0               NaN                NaN                NaN                NaN   
1               NaN                NaN                NaN                NaN   

   Post Hashtag Clicks  Post Detail Expand Clicks  Profile Clicks  \
0                  NaN                        NaN             NaN   
1                  NaN                        NaN             NaN   

   Post Photo View Clicks  Post Video Play Clicks  Other Engagements  Answers  \
0                     NaN                     NaN                NaN      NaN   
1                     NaN                     NaN                NaN      NaN   

   App Engagements  App Install Attempts  App Opens  Follows from Post  \
0              NaN                   NaN        NaN                NaN   
1         

## Data Cleaning

### Drop Columns

The plan is to drop columns in which atleast 80% of the data are null values.

These columns cannot be used in the analysis

In [5]:
# Return columns where null values are >= 80%
null_columns = list(df.columns[df.isnull().sum() >= 0.8*len(df)])

In [6]:
# Get the columns that are not completely null
not_null_columns = df.columns[~df.columns.isin(null_columns)]

# Create a new dataframe with only the non-null columns
df_not_null = df[not_null_columns]

In [7]:
# Check new dataframe dimensions
print('Rows: ', df_not_null.shape[0])
print('Columns: ', df_not_null.shape[1])

Rows:  10000
Columns:  20


In [8]:
# Examine first two observations
df_not_null.head(2)

Date            Post ID    Network Post Type Content Type  \
0  1/30/2023 5:06 pm  18216468940204235  Instagram      Post     Carousel   
1  4/15/2022 9:04 am  17971531720518163  Instagram      Post        Photo   

       Profile        Sent by                                      Link  \
0  stanbicibtc                 https://www.instagram.com/p/CoC6CKBIHbV/   
1  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CcXUQTEMRm7/   

                                                                                                                                                                                                                                                                           Post  \
0  The new Naira banknotes in circulation have several security features for ease of recognition which can be easily seen and touched.\nPlease deposit your old N200, N500 and N1000 naira notes at any Stanbic IBTC branch near you.\nImage Source - @centralbankng \n#ITCANBE   
1                                                                                                                                                                       We hope the faith that this #GoodFriday restores abides with you and your loved ones. \n#TGIF\n#ITCANBE   

  Impressions Organic Impressions   Reach Organic Reach  \
0      39,681              39,681  32,418        32,418   
1      29,186              29,186  25,055        25,055   

  Engagement Rate (per Impression) Engagement Rate (per Reach) Engagements  \
0                            4.89%                       5.99%       1,942   
1                            6.04%                       7.04%       1,763   

  Reactions  Likes Comments  Saves  
0     1,261  1,261       56  625.0  
1       709    709    1,033   21.0

In [9]:
# print column names 
print(df_not_null.columns)

Index(['Date', 'Post ID', 'Network', 'Post Type', 'Content Type', 'Profile',
       'Sent by', 'Link', 'Post', 'Impressions', 'Organic Impressions',
       'Reach', 'Organic Reach', 'Engagement Rate (per Impression)',
       'Engagement Rate (per Reach)', 'Engagements', 'Reactions', 'Likes',
       'Comments', 'Saves'],
      dtype='object')


### Quality Issues 

The plan is to examine the data for
* duplicates
* wrong data types
* inconsistent / wrong formatting and
* missing data

In [10]:
# Check duplicates
print('Number of duplicates: ', df_not_null.duplicated().sum())

Number of duplicates:  0


In [11]:
# Examine datatypes
pd.DataFrame(df_not_null.dtypes)

0
Date                               object
Post ID                             int64
Network                            object
Post Type                          object
Content Type                       object
Profile                            object
Sent by                            object
Link                               object
Post                               object
Impressions                        object
Organic Impressions                object
Reach                              object
Organic Reach                      object
Engagement Rate (per Impression)   object
Engagement Rate (per Reach)        object
Engagements                        object
Reactions                          object
Likes                              object
Comments                           object
Saves                             float64

In [12]:
# Check missing values
pd.DataFrame(df_not_null.isna().sum()).sort_values(by=0, ascending=False)

0
Engagement Rate (per Reach)       2747
Engagement Rate (per Impression)  1768
Organic Impressions               1484
Impressions                       1484
Organic Reach                     1484
Reach                             1484
Saves                             1484
Post                                11
Link                                 0
Sent by                              0
Post ID                              0
Profile                              0
Content Type                         0
Post Type                            0
Network                              0
Engagements                          0
Reactions                            0
Likes                                0
Comments                             0
Date                                 0

#### List of Quality Issues

* **8** columns have missing values
* 'Engagement Rate (per Impression)' & 'Engagement Rate (per Reach)'  columns have % sign values
* Several columns have the wrong datatype
    * 10 columns are wrongly labeled as strings instead of numerics.
    * The Date column should be a datetime object.
    * 'Sent by' and 'Content Type' columns should be categorical.
* Several columns have values are formatted with ',' e.g. '200,000'. This format cannot be converted to numeric without issues. They must be treated while converting datatypes.

* 'Sent by' column has blank values (i.e missing values are filled with space).


---

#### Fixing Missing Data Issues

In [13]:
# Check rows where Impressions are null.
df_not_null[df_not_null['Impressions'].isna()].head()

Date            Post ID    Network Post Type Content Type  \
8516  6/12/2023 1:08 pm  18029321227525834  Instagram      Post     Carousel   
8517   6/8/2023 9:55 am  18365455210048288  Instagram      Post        Photo   
8518   6/7/2023 1:30 pm  17964499847327992  Instagram      Post        Photo   
8519  5/16/2023 8:30 pm  18005669605736790  Instagram      Post        Photo   
8520  4/27/2023 8:18 pm  17979979232158293  Instagram      Post        Photo   

          Profile        Sent by                                      Link  \
8516  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CtY8c1Ts0pP/   
8517  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CtOTPwcMPZ-/   
8518  stanbicibtc      Kemi Amoo  https://www.instagram.com/p/CtMG-nLOyVP/   
8519  stanbicibtc      Kemi Amoo  https://www.instagram.com/p/CsUNo24OrlI/   
8520  stanbicibtc   Sophia Amuka  https://www.instagram.com/p/CrjRFyLByBP/   

                                                                                                                                                                                                                                                                                                                                                                                                                                 Post  \
8516  At Stanbic IBTC, we hold special gatherings for the Chinese community in order to better understand their needs and give tailored financial solutions. Customers who partner with us get access to a wide range of services and benefits that are tailored to their specific financial needs.​\nFor further information, please visit our website https://bit.ly/EnterpriseAcademySolutions ​\n\n#ITCANBE​\n#EnterpriseAcademy​   
8517                                                                                                                                                                                                           Motor insurance is required for any ‘Idan’ who owns a vehicle, it protects them financially from potential losses and liabilities. Send an email to CallMe@stanbicibtc.com today to safeguard your asset.​\n\n#ITCANBE   
8518                                                                                                                                                                                                                                                                       Did you know that you can use your retirement savings account to fund your home loan? Join us on our YouTube Live and we'll walk you through it.\n#ITCANBE   
8519                                                                                                                                                                                                                                                       We celebrate the boy child today and every day for their unique potential. May he be empowered to reach for the stars and beyond.\n\n#InternationalDayofBoyChild\n#ITCANBE   
8520                                                                                                                                                            Make your money bloom with the right investments. Get all the insights you need from this Fireside Chat at the Bloom Conference 1.0.\nVisit the link in our bio and click 'Bloom Conference' to download the Stanbic IBTC Events app to register.\n#GoBloom\n#ITCANBE   

     Impressions Organic Impressions Reach Organic Reach  \
8516         NaN                 NaN   NaN           NaN   
8517         NaN                 NaN   NaN           NaN   
8518         NaN                 NaN   NaN           NaN   
8519         NaN                 NaN   NaN           NaN   
8520         NaN                 NaN   NaN           NaN   

     Engagement Rate (per Impression) Engagement Rate (per Reach) Engagements  \
8516                              NaN                         NaN           

##### Note

5 columns have missing data corresponding to deleted posts (1484 observations).

These rows will be deleted.

In [14]:
# Remove rows where Impressions are null.
df_not_null = df_not_null[~df_not_null['Impressions'].isna()]

---

In [15]:
# Check Engagements, Impressions, and Reach where Engagement Rate (per Reach) is null.   
(df_not_null[df_not_null['Engagement Rate (per Impression)'].isna()]
[['Engagements', 'Impressions', 'Reach', 'Engagement Rate (per Impression)', 'Engagement Rate (per Reach)']])

Engagements Impressions Reach Engagement Rate (per Impression)  \
7248           5           0     0                              NaN   
7249           5           0     0                              NaN   
7251           2           0     0                              NaN   
7252           2           0     0                              NaN   
7253          11           0     0                              NaN   
...          ...         ...   ...                              ...   
8511          64           0     0                              NaN   
8512          50           0     0                              NaN   
8513          59           0     0                              NaN   
8514          52           0     0                              NaN   
8515         102           0     0                              NaN   

     Engagement Rate (per Reach)  
7248                         NaN  
7249                         NaN  
7251                         NaN  
7252                         NaN  
7253                         NaN  
...                          ...  
8511                         NaN  
8512                         NaN  
8513                         NaN  
8514                         NaN  
8515                         NaN  

[284 rows x 5 columns]

##### Note

Missing values in Engagement Rate (per Impression) and Engagement Rate (per Reach) correspond to 0 Impressions or Reach

These values will be imputed with 0

In [16]:
# Impute 0 for missing values in 'Engagement Rate (per Impression)' and 'Engagement Rate (per Reach)' columns
df_not_null['Engagement Rate (per Impression)'] = df_not_null['Engagement Rate (per Impression)'].fillna(0)
df_not_null['Engagement Rate (per Reach)'] = df_not_null['Engagement Rate (per Reach)'].fillna(0)

In [17]:
# Confirm imputation
df_not_null[['Engagement Rate (per Impression)', 'Engagement Rate (per Reach)']].isna().sum()

Engagement Rate (per Impression)    0
Engagement Rate (per Reach)         0
dtype: int64

---

In [18]:
# Examine missing posts
df_not_null[df_not_null['Post'].isna()][['Post', 'Link', 'Content Type']]

Post                                       Link Content Type
1423  NaN  https://www.instagram.com/tv/CUsLOdsoeK7/        Video
2687  NaN  https://www.instagram.com/tv/Ckqad7iI4Lx/        Video
3037  NaN  https://www.instagram.com/tv/CPnqTksg0xY/        Video
3466  NaN   https://www.instagram.com/p/B4rgFtLFS0Y/        Photo
4057  NaN  https://www.instagram.com/tv/CSMdKslodw-/        Video
4680  NaN  https://www.instagram.com/tv/CXOWnewoDbO/        Video
4721  NaN  https://www.instagram.com/tv/CWGPHnaIN6D/        Video
4881  NaN  https://www.instagram.com/tv/CXlm5xEoRdP/        Video
5297  NaN  https://www.instagram.com/tv/CXlmQIroVuz/        Video

In [19]:
# Examine content type of missing posts
df_not_null[df_not_null['Post'].isna()][['Content Type']].value_counts()

Content Type
Video           8
Photo           1
dtype: int64

##### Note

Posts with missing values do not have text content

They are all media content (especially videos)

These values will be imputed with blank space instead.

In [20]:
# fill missing values in 'Post' column with values in ''
df_not_null['Post'] = df_not_null['Post'].fillna('')

In [21]:
# Confirm imputation
df_not_null['Post'].isna().sum()

0

In [22]:
# Check Impressions and 'Organic Impressions' for differences
df_not_null[(df_not_null['Impressions'].notna()) &
    (df_not_null['Impressions'] != df_not_null['Organic Impressions'])
            ][['Impressions', 'Organic Impressions']]

Empty DataFrame
Columns: [Impressions, Organic Impressions]
Index: []

----

##### Additional columns to be dropped


* **'Post ID'**: There is no reason to track individual posts in this analysis.
* **Network, 'Post Type', 'Profile**: Redundant Information
* **Link**: The actual link to instagram posts are not needed for analysis
* **Organic Impressions**: This column has the same data as the Impression column

In [23]:
df_not_null.columns

Index(['Date', 'Post ID', 'Network', 'Post Type', 'Content Type', 'Profile',
       'Sent by', 'Link', 'Post', 'Impressions', 'Organic Impressions',
       'Reach', 'Organic Reach', 'Engagement Rate (per Impression)',
       'Engagement Rate (per Reach)', 'Engagements', 'Reactions', 'Likes',
       'Comments', 'Saves'],
      dtype='object')

In [24]:
# Drop additional columns 
cols = ['Post ID', 'Network', 'Post Type', 'Profile','Link', 'Organic Impressions']

df_not_null = df_not_null.drop(cols, axis=1)

In [25]:
# Check missing values
pd.DataFrame(df_not_null.isna().sum()).sort_values(by=0, ascending=False)

0
Date                              0
Content Type                      0
Sent by                           0
Post                              0
Impressions                       0
Reach                             0
Organic Reach                     0
Engagement Rate (per Impression)  0
Engagement Rate (per Reach)       0
Engagements                       0
Reactions                         0
Likes                             0
Comments                          0
Saves                             0

---
#### Fix issues with 'Engagement Rate (per impression)', 'Engagement Rate (per Reach)' Formatting

In [26]:
# Remove % sign from 'Engagement Rate (per Impression)', 'Engagement Rate (per Reach)' & 'Click-Through Rate' columns
df_not_null['Engagement Rate (per Impression)'] = df_not_null['Engagement Rate (per Impression)'].str.replace('%', '')
df_not_null['Engagement Rate (per Reach)'] = df_not_null['Engagement Rate (per Reach)'].str.replace('%', '')

In [27]:
# Confirm changes
df_not_null[['Engagement Rate (per Impression)', 'Engagement Rate (per Reach)']].head()

Engagement Rate (per Impression) Engagement Rate (per Reach)
0                             4.89                        5.99
1                             6.04                        7.04
2                             3.48                        3.83
3                             4.48                        5.53
4                             6.58                        7.54

---
#### Fix issues with wrong data types

In [28]:
# Convert "Date" column to datetime
df_not_null['Date'] = pd.to_datetime(df_not_null['Date'])

In [29]:
# List of columns to convert to numeric
numeric_columns =  ['Impressions', 'Reach', 'Organic Reach', 'Engagement Rate (per Impression)',
       'Engagement Rate (per Reach)', 'Engagements', 'Reactions', 'Likes', 'Comments']

In [30]:
# Convert columns to numeric data type
df_not_null[numeric_columns] = df_not_null[numeric_columns].applymap(lambda x: float(str(x).replace(',', '')))

In [31]:
# Convert 'Sent by' and 'Content type' columns to categorical data type
df_not_null['Sent by'] = df_not_null['Sent by'].astype('category')
df_not_null['Content Type'] = df_not_null['Content Type'].astype('category')

In [32]:
# Examine datatypes
pd.DataFrame(df_not_null.dtypes)

0
Date                              datetime64[ns]
Content Type                            category
Sent by                                 category
Post                                      object
Impressions                              float64
Reach                                    float64
Organic Reach                            float64
Engagement Rate (per Impression)         float64
Engagement Rate (per Reach)              float64
Engagements                              float64
Reactions                                float64
Likes                                    float64
Comments                                 float64
Saves                                    float64

---

#### Fix issues with 'Sent by' blank spaces

In [36]:
pd.DataFrame(df_not_null['Sent by'].value_counts())

Sent by
Other                 3544
Blessing Ubah         2668
Aramide Salami        1717
Kemi Amoo              311
Sophia Amuka           190
Damilare Oyekanmi       86

In [34]:
# Replace blank space with 'Other' in 'Sent by' column
df_not_null['Sent by'] = df_not_null['Sent by'].replace(' ', 'Other')

In [35]:
# Confirm changes
pd.DataFrame(df_not_null['Sent by'].value_counts())

Sent by
Other                 3544
Blessing Ubah         2668
Aramide Salami        1717
Kemi Amoo              311
Sophia Amuka           190
Damilare Oyekanmi       86

---